pkg/server/image_pull.go

```diff
@@ -88,70 +81,66 @@ func (c *criContainerdService) PullImage(ctx context.Context, r *runtime.PullIma
 				r.GetImage().GetImage(), retRes.GetImageRef())
 		}
 	}()
+
 	imageRef := r.GetImage().GetImage()
+	namedRef, err := normalizeImageRef(imageRef)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse image reference %q: %v", imageRef, err)
+	}
+	// TODO(random-liu): [P0] Avoid concurrent pulling/removing on the same image reference.
+	ref := namedRef.String()
+	if ref != imageRef {
+		glog.V(4).Infof("PullImage using normalized image ref: %q", ref)
+	}
+
+	resolver := docker.NewResolver(docker.ResolverOptions{
+		Credentials: func(string) (string, string, error) { return ParseAuth(r.GetAuth()) },
+		Client:      http.DefaultClient,
+	})
 
 	// TODO(mikebrow): add truncIndex for image id
-	imageID, repoTag, repoDigest, err := c.pullImage(ctx, imageRef, r.GetAuth())
+	image, err := c.client.Pull(ctx, ref, containerd.WithPullUnpack, containerd.WithSchema1Conversion, containerd.WithResolver(resolver))
 	if err != nil {
-		return nil, fmt.Errorf("failed to pull image %q: %v", imageRef, err)
+		return nil, fmt.Errorf("failed to pull image %q: %v", ref, err)
 	}
-	glog.V(4).Infof("Pulled image %q with image id %q, repo tag %q, repo digest %q", imageRef, imageID,
+	repoDigest, repoTag := getRepoDigestAndTag(namedRef, image.Target().Digest, image.Target().MediaType == containerdimages.MediaTypeDockerSchema1Manifest)
+	for _, r := range []string{repoTag, repoDigest} {
```

Comment:
Got it . Thanks for explaining. Is there anyway we could avoid storing these many-to-one reference copies in the store ? I mean any change we can do on the container to resolve the image ? Not in the scope of this PR. But moving forward.